src/object_store/http.rs

```diff
@@ -0,0 +1,385 @@
+    ) -> object_store::Result<BoxStream<'_, object_store::Result<ObjectMeta>>> {
+        // DataFusion uses the HEAD request instead of it's listing a single file
+        // (path doesn't end with a slash). Since HTTP doesn't support listing anyway,
+        // this makes our job easier.
+
+        match prefix {
+            None => Err(HttpObjectStoreError::ListingUnsupported.into()),
+            Some(p) => {
+                let p_str = p.to_string();
+                if p_str.ends_with('/') {
+                    Err(HttpObjectStoreError::ListingUnsupported.into())
+                } else {
+                    // Use the HEAD implementation instead
+                    Ok(Box::pin(stream::iter(vec![self.head(p).await])))
+                }
+            }
+        }
+    }
+
+    async fn list_with_delimiter(
+        &self,
+        _prefix: Option<&Path>,
+    ) -> object_store::Result<ListResult> {
+        // Not used by DF, so we punt on implementing this
+        Err(object_store::Error::NotImplemented)
+    }
+
+    async fn copy(&self, _from: &Path, _to: &Path) -> object_store::Result<()> {
+        Err(object_store::Error::NotSupported {
+            source: Box::new(HttpObjectStoreError::WritesUnsupported),
+        })
+    }
+
+    async fn copy_if_not_exists(
+        &self,
+        _from: &Path,
+        _to: &Path,
+    ) -> object_store::Result<()> {
+        Err(object_store::Error::NotSupported {
+            source: Box::new(HttpObjectStoreError::WritesUnsupported),
+        })
+    }
+}
+
+/// Add HTTP/HTTPS support to a DataFusion SessionContext
+pub fn add_http_object_store(context: &SessionContext) {
+    context.runtime_env().register_object_store(
+        "http",
+        ANYHOST,
+        Arc::new(HttpObjectStore::new("http".to_string())),
+    );
+
+    context.runtime_env().register_object_store(
+        "https",
+        ANYHOST,
+        Arc::new(HttpObjectStore::new("https".to_string())),
+    );
+}
+
+/// Prepare a URL (`LOCATION` clause) so that DataFusion can directly route it to the object store.
+///
+/// This is done by injecting a special `anyhost` hostname inside of the URL, for example:
+/// `https://some-host.com/file.parquet` -> `https://anyhost/some-host.com/file.parquet`. This is
+/// because DataFusion routes URLs to object store instances based on the scheme and the host. Hence,
+/// in this case, `LOCATION 'https://some-host.com/file.parquet'` will get DataFusion to
+/// pass the path to us as `some-host.com/file.parquet` (as it strips the scheme and the fake host)
+/// and the object store instance will re-append the correct scheme back to the URL.
+///
+/// Returns a `None` if the location doesn't start with `http://` / `https://`.
+pub fn try_prepare_http_url(location: &str) -> Option<String> {
+    location.strip_prefix("http://").map_or_else(
+        || {
+            location
+                .strip_prefix("https://")
+                .map(|l| format!("https://{}/{}", ANYHOST, l))
```

Comment:
Note: An IPv6 host must be surrounded by square brackets when serialized to a URI (unless there is no port specified, in some cases?). This code looks correct, as far as I can tell.

I just felt it was worth the warning, because yesterday I spent time I will never get back on tracing this exact bug in multiple JS libraries that incorrectly serialized a URI with an IPv6 host like `::1`. (It's a new problem because Node 18 now binds to `localhost` using the resolution order of the operating system, meaning code can no longer expect `server.address()` to return an IPv4 address, even for `localhost`).

So basically I recommend adding some tests for a URI with `::1` as the host 😸

Reply:
Thanks for the warning. We don't serialize/deserialize IPs here, we just move the existing URL around, as passed by the user. If they don't know how to serialize IPs properly, it's their fault :)